timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py

```diff
@@ -634,3 +634,131 @@ def get_reindexed_view(self, freq: str = "S") -> TimeSeriesDataFrame:
+
+        Parameters
+        ----------
+        freq: str
+            Frequency string of the new time series data index.
+
+        Examples
+        --------
+        >>> print(ts_dataframe)
+                            target
+        item_id timestamp
+        0       2019-01-01     NaN
+                2019-01-03     1.0
+                2019-01-06     2.0
+                2019-01-07     NaN
+        1       2019-02-04     3.0
+                2019-02-07     4.0
+
+        >>> print(ts_dataframe.to_regular_index(freq="D"))
+                            target
+        item_id timestamp
+        0       2019-01-01     NaN
+                2019-01-02     NaN
+                2019-01-03     1.0
+                2019-01-04     NaN
+                2019-01-05     NaN
+                2019-01-06     2.0
+                2019-01-07     NaN
+        1       2019-02-04     3.0
+                2019-02-05     NaN
+                2019-02-06     NaN
+                2019-02-07     4.0
+
+        """
+        if self.freq is not None:
+            if self.freq != freq:
+                raise ValueError(
+                    f"TimeSeriesDataFrame already has a regular index with freq '{self.freq}' "
+                    f"that cannot be converted to the given freq '{freq}'"
+                )
+            else:
+                return self
+
+        filled_series = []
+        for item_id, time_series in self.groupby(level=ITEMID, sort=False):
+            time_series = time_series.droplevel(ITEMID)
+            timestamps = time_series.index
+            if not timestamps.is_monotonic_increasing:
+                raise ValueError(
+                    "Please make sure that the timestamps are sorted in increasing order for all time series."
+                )
+            start = timestamps[0]
+            end = timestamps[-1]
+            filled_timestamps = pd.date_range(start=start, end=end, freq=freq, name=TIMESTAMP)
+            if not timestamps.isin(filled_timestamps).all():
+                raise ValueError(
+                    f"Irregularly-sampled timestamps in this TimeSeriesDataFrame are not compatible "
+                    f"with the given frequency '{freq}'"
+                )
+            filled_series.append(pd.concat({item_id: time_series.reindex(filled_timestamps)}, names=[ITEMID]))
+
+        return TimeSeriesDataFrame(pd.concat(filled_series), static_features=self.static_features)
+
+    def fill_missing_values(self, method: str = "ffill") -> "TimeSeriesDataFrame":
+        """Fill missing values represented by NaN.
+
+        Missing values at the beginning of each time series are dropped. Remaining missing values (in the middle and
+        at the end) are filled using the chosen method.
+
+
+        Parameters
+        ----------
+        method : {"ffill", "interpolate"}, default = "ffill"
+            Method used to impute missing values.
+            "ffill" - propagate last valid observation forward.
```

Comment:
We should support all the filling [methods](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html#pandas.DataFrame.fillna) available in pandas.

Reply:
I see two potential problems with `bfill`/`backfill`:
- it doesn't fill the trailing NaNs, and these are the ones that are actually important to ensure that all models can generate predictions over the forecast horizon. E.g, after we bfill `[1, 1, NaN, NaN]`, we again get `[1, 1, NaN, NaN]` and this cannot be processed by `TimeSeriesPredictor`. We cannot just drop the trailing NaNs as easily as we can drop leading NaNs with `ffill`.
- it introduces information leakage from the test/val set, which might affect model selection.

Do you think there is a strong potential use case for `bfill` that we should support?

Reply:
Information leakage is a critical point, agree that we should drop `bfill` for now. Curious for interpolate, will linear interpolation cause leakage as well by any chance?

Reply:
Good point regarding `interpolation`. I've checked how other libraries ([sktime](https://www.sktime.org/en/v0.16.0/api_reference/auto_generated/sktime.transformations.series.impute.Imputer.html), [darts](https://unit8co.github.io/darts/_modules/darts/utils/missing_values.html#fill_missing_values)) handle this and updated the functionality to be more in line with them. 
- Do not change the index (never drop the leading NaNs)
- By default, use `ffill` to fill gaps + trailing NaNs, then use `bfill` to fill the leading NaNs.
- Added options `constant` and `bfill`.
- Added warnings for `bfill` and `interpolate` that these may lead to data leakage.